aws-s3-transfer-manager/src/operation/download_objects/handle.rs

```diff
@@ -19,14 +21,57 @@ pub struct DownloadObjectsHandle {
 
 impl DownloadObjectsHandle {
     /// Consume the handle and wait for download transfer to complete
+    ///
+    /// When the `FailedTransferPolicy` is set to [`FailedTransferPolicy::Abort`], this method
+    /// will return the first error if any of the spawned tasks encounter one. The other tasks
+    /// will be canceled, but their cancellations will not be reported as errors by this method;
+    /// they will be logged as errors, instead.
+    ///
+    /// If the `FailedTransferPolicy` is set to [`FailedTransferPolicy::Continue`], the
+    /// [`DownloadObjectsOutput`] will include a detailed breakdown, such as the number of
+    /// successful downloads and the number of failed ones.
+    ///
+    // TODO(aws-sdk-rust#1159) - Consider if we want to return other all errors encountered during cancellation.
     #[tracing::instrument(skip_all, level = "debug", name = "join-download-objects")]
     pub async fn join(mut self) -> Result<DownloadObjectsOutput, crate::error::Error> {
-        // TODO - Consider implementing more sophisticated error handling such as canceling in-progress transfers
+        let mut first_error_to_report = None;
         // join all tasks
         while let Some(join_result) = self.tasks.join_next().await {
-            join_result??;
+            let result = join_result.expect("task completed");
+            if let Err(e) = result {
+                match self.ctx.state.input.failure_policy() {
+                    FailedTransferPolicy::Abort
+                        if first_error_to_report.is_none()
+                            && e.kind() != &ErrorKind::OperationCancelled =>
+                    {
+                        first_error_to_report = Some(e);
+                    }
+                    FailedTransferPolicy::Continue => {
+                        tracing::warn!("encountered but dismissed error when the failure policy is `Continue`: {e}")
```

Comment:
Looks like we already do to some degree, like [here](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/06c087a5d53676bb048f6c512b8eb1fda63f03d5/aws-s3-transfer-manager/src/operation/download_objects/worker.rs#L102-L106) and [here](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/06c087a5d53676bb048f6c512b8eb1fda63f03d5/aws-s3-transfer-manager/src/operation/upload_objects/worker.rs#L218) regardless of the failure policy. The above location at line 50 is a bit of an edge case because with `Continue`, any error is supposed to be dismissed and the execution continues, so by the time the execution flow reaches line 50, we probably do not have access to an error instance that has useful info like the object key.

That said, maybe you could point me to an example location that could benefit additional logging. Happy to add it in a subsequent PR.